### ISSUE TITLE
Conditions do not work (sometimes)

### ISSUE THREAD

**Issue by ShlomiPorush** [nielsfaber/scheduler-component#164]
### Checklist
- [ V] I checked for similar existing issues (both open and closed) before posting.
- [ V] I am running the latest version.

### Expected behavior
Did something change in recent updates? conditions just do not work
A simple schedule to change my living room shutter with 2 conditions to work only on Saturday when we are at home
So 1 boolean for if we are at home - changes manually, and one more boolean for Saturday and holidays. 

Enabling 'Re-evaluate when conditions change' did not help. 

![1](https://user-images.githubusercontent.com/5992161/147950002-22c8f38d-4c60-481e-a964-c864ea4925a0.PNG)
![2](https://user-images.githubusercontent.com/5992161/147950010-b3e45362-d387-4b56-996b-ae3231d74a45.PNG)

Other schedules (with AC & lights) with the same conditions working fine.

All is up to date.

the schedule to run only when conditions are met.

### Actual behavior
the schedule ignores the conditions.

### Steps to Reproduce
the

**Comment by nielsfaber** [comment]
I will take a look at this somewhere in the next days.
A 'sometimes' problem is usually harder to catch and to fix.
If you have extra clues for me, please let me know. 
Does it also occur to you if a schedule has only 1 condition?

**Comment by ShlomiPorush** [comment]
I completely forgot about it, I stopped using scheduler. 
Just checked again & I can confirm with just 1 condition it doesn't happen. With 2 conditions it happens all the time not just sometimes.. All the time.

**Comment by nielsfaber** [comment]
To test this issue, I made a schedule which turns on an input_boolean 'scheduler-test':
![image](https://user-images.githubusercontent.com/16861605/150067063-e82a721f-61ae-4c12-9a2d-3bd30f7bee51.png)

The schedule has condition that 'condition A' and 'condition B' need to be on (both are input_booleans as well):
![image](https://user-images.githubusercontent.com/16861605/150067158-86042b71-086d-4f63-a816-f226990b1377.png)

I made a card with the condition booleans + 'schedule test' booleans to test:
![image](https://user-images.githubusercontent.com/16861605/150067233-e6098f95-6447-48f8-9418-c73bbe730cd1.png)

For me 'schedule test' switch will only turn on if the 'condition A' and 'condition B' booleans are turned on.
This is the behaviour as I would expect it, so for me the problem does not occur.

I can repeat the test 10 times, with the same effect. So I cannot reproduce your problem.
Perhaps you could enable debug logging for scheduler to get more info on the problem?
You can do so by adding to configuration.yaml:
```
logger:
  default: warning
  logs:
    custom_components.scheduler: debug
```

Debug logging when only one condition is met:
> [custom_components.scheduler.actions] [b7c209]: State of input_boolean.condition_a has changed, re-evaluating actions
[custom_components.scheduler.actions] [b7c209]: Conditions have failed, skipping execution of actions

Debug logging when both conditions are met:
> [custom_components.scheduler.actions] [b7c209]: State of input_boolean.condition_b has changed, re-evaluating actions
[custom_components.scheduler.actions] [b7c209]: Executing service input_boolean.turn_on on entity input_boolean.scheduler_test

Also if you could share some data from 'history' when the problem occurs, that would be useful.
At this point I don't know how to help you, I cannot fix a problem which I cannot observe.

**Comment by ShlomiPorush** [comment]
Thank you for the time you took to check it out. So I replicated it in my setup and it works wonderfully.
2 conditions, 1 needs to be off & one needs to be on... works as expected. 

This got me wondering why it is different from the schedule I have, so for the next test, I enabled debugging. 
I created a new schedule for my living room shutter with 2 conditions. 1 needs to be on & the other one needs to be off... worked great. so whats is different between the new schedule VS the existed schedule.

debug log from the schedule that is NOT working and ignores the conditions.

`2022-01-19 11:04:55 DEBUG (MainThread) [custom_components.scheduler.timer] Timer of 7a9cc9 set for 2022-01-19 13:00:00+02:00
2022-01-19 11:04:55 DEBUG (MainThread) [custom_components.scheduler.switch] Schedule 7a9cc9 is starting in a timeslot, proceed with actions
2022-01-19 11:04:55 DEBUG (MainThread) [custom_components.scheduler.actions] [7a9cc9]: Executing service cover.set_cover_position on entity cover.living_room_shutter
2022-01-19 11:04:55 DEBUG (MainThread) [custom_components.scheduler.actions] [7a9cc9]: Finished execution of actions`
![12](https://user-images.githubusercontent.com/5992161/150104122-fffbe97c-998b-4632-b2c3-a0e032512cd4.png)

**Comment by nielsfaber** [comment]
From your loggings it shows that your original schedule (Schedule 7a9cc9) does not check for conditions, so it looks like no conditions are assigned to that schedule.
There could be an issue with the saving of the schedule.

Maybe its good to take a look in the `.storage` folder (should be inside your HA folder where configuration.yaml etc are) in the `scheduler.storage` file. Here you can find the complete configuration for item `7a9cc9`.
Please share this section.

**Comment by ShlomiPorush** [comment]
Sure. 

`            {
                "schedule_id": "7a9cc9",
                "timeslots": [
                    {
                        "start": "00:00:00",
                        "stop": "09:00:00",
                        "conditions": [
                            {
                                "entity_id": "input_boolean.sabbath_mode",
                                "attribute": "state",
                                "value": "on",
                                "match_type": "is"
                            },
                            {
                                "entity_id": "input_boolean.sabbath_away",
                                "attribute": "state",
                                "value": "off",
                                "match_type": "is"
                            }
                        ],
                        "condition_type": "and",
                        "track_conditions": false,
                        "actions": [
                            {
                                "service": "cover.close_cover",
                                "entity_id": "cover.living_room_shutter",
                                "service_data": {}
                            }
                        ]
                    },
                    {
                        "start": "09:00:00",
                        "stop": "13:00:00",
                        "conditions": [],
                        "condition_type": null,
                        "track_conditions": false,
                        "actions": [
                            {
                                "service": "cover.set_cover_position",
                                "entity_id": "cover.living_room_shutter",
                                "service_data": {
                                    "position": 30
                                }
                            }
                        ]
                    },
                    {
                        "start": "13:00:00",
                        "stop": "16:30:00",
                        "conditions": [],
                        "condition_type": null,
                        "track_conditions": false,
                        "actions": [
                            {
                                "service": "cover.set_cover_position",
                                "entity_id": "cover.living_room_shutter",
                                "service_data": {
                                    "position": 25
                                }
                            }
                        ]
                    },
                    {
                        "start": "16:30:00",
                        "stop": "21:50:00",
                        "conditions": [],
                        "condition_type": null,
                        "track_conditions": false,
                        "actions": [
                            {
                                "service": "cover.set_cover_position",
                                "entity_id": "cover.living_room_shutter",
                                "service_data": {
                                    "position": 23
                                }
                            }
                        ]
                    },
                    {
                        "start": "21:50:00",
                        "stop": "00:00:00",
                        "conditions": [],
                        "condition_type": null,
                        "track_conditions": false,
                        "actions": [
                            {
                                "service": "cover.close_cover",
                                "entity_id": "cover.living_room_shutter",
                                "service_data": {}
                            }
                        ]
                    }
                ],
                "weekdays": [
                    "daily"
                ],
                "start_date": null,
                "end_date": null,
                "repeat_type": "repeat",
                "name": "Sabbath Living Room Shutter",
                "enabled": false
            },`

**Comment by ShlomiPorush** [comment]
and here is the config for the new schedule I setup today

                "schedule_id": "698c78",
                "timeslots": [
                    {
                        "start": "00:00:00",
                        "stop": "08:00:00",
                        "conditions": [
                            {
                                "entity_id": "input_boolean.sabbath_mode",
                                "attribute": "state",
                                "value": "on",
                                "match_type": "is"
                            },
                            {
                                "entity_id": "input_boolean.sabbath_away",
                                "attribute": "state",
                                "value": "off",
                                "match_type": "is"
                            }
                        ],
                        "condition_type": "and",
                        "track_conditions": false,
                        "actions": [
                            {
                                "service": "cover.close_cover",
                                "entity_id": "cover.living_room_shutter",
                                "service_data": {}
                            }
                        ]
                    },
                    {
                        "start": "08:00:00",
                        "stop": "16:00:00",
                        "conditions": [
                            {
                                "entity_id": "input_boolean.sabbath_mode",
                                "attribute": "state",
                                "value": "on",
                                "match_type": "is"
                            },
                            {
                                "entity_id": "input_boolean.sabbath_away",
                                "attribute": "state",
                                "value": "off",
                                "match_type": "is"
                            }
                        ],
                        "condition_type": "and",
                        "track_conditions": false,
                        "actions": [
                            {
                                "service": "cover.set_cover_position",
                                "entity_id": "cover.living_room_shutter",
                                "service_data": {
                                    "position": 20
                                }
                            }
                        ]
                    },
                    {
                        "start": "16:00:00",
                        "stop": "20:00:00",
                        "conditions": [
                            {
                                "entity_id": "input_boolean.sabbath_mode",
                                "attribute": "state",
                                "value": "on",
                                "match_type": "is"
                            },
                            {
                                "entity_id": "input_boolean.sabbath_away",
                                "attribute": "state",
                                "value": "off",
                                "match_type": "is"
                            }
                        ],
                        "condition_type": "and",
                        "track_conditions": false,
                        "actions": [
                            {
                                "service": "cover.close_cover",
                                "entity_id": "cover.living_room_shutter",
                                "service_data": {}
                            }
                        ]
                    },
                    {
                        "start": "20:00:00",
                        "stop": "00:00:00",
                        "conditions": [],
                        "condition_type": null,
                        "track_conditions": false,
                        "actions": [
                            {
                                "service": "cover.set_cover_position",
                                "entity_id": "cover.living_room_shutter",
                                "service_data": {
                                    "position": 13
                                }
                            }
                        ]
                    }
                ],
                "weekdays": [
                    "daily"
                ],
                "start_date": null,
                "end_date": null,
                "repeat_type": "repeat",
                "name": "",
                "enabled": false
            },

**Comment by nielsfaber** [comment]
Bingo, there's your problem.

You can see that the conditions are only stored in the timeslot from 00:00-09:00, the remaining timeslots have an empty conditions list.
There used to be a problem which caused this, but it has been fixed in June last year: https://github.com/nielsfaber/scheduler-card/issues/415

Could your schedule be from before that time, or it is newer?
In the second case, there might be some other problem in the card that we need to investigate. 
In the first case, the schedule should be 'repaired', but I'm not sure if this is done by editing + saving it in the card.

**Comment by ShlomiPorush** [comment]
I created the schedule ID 7a9cc9 about a month ago.
look at the schedule 698c78 I created today, from my understanding I hours between 20:00 - 00:00 also has no conditions...  

> "start": "20:00:00",
> "stop": "00:00:00",
> "conditions": [],

**Comment by nielsfaber** [comment]
In that case there is still a problem in the card.
I will look into it.

**Comment by szatan81** [comment]
I can confirm the same situation using just 1 condition. With 10 schedules created from the card one has missing conditions in scheduler.storage

**Comment by nielsfaber** [comment]
The problem was caused by the card when clicking the 'add' button. The newly added slot would not have the previously defined conditions stored. It is fixed now: https://github.com/nielsfaber/scheduler-card/commit/f45fdf67b56612ce8c75b84b90dd9a662e30f271# 

I also made a fix in the scheduler-component which will 'repair' all schedules by checking which have multiple timeslots with different number of conditions assigned, and copying them from the first slot: https://github.com/nielsfaber/scheduler-component/commit/652e3eb9e1b586ca5bc6899b84052dd15ae41fac

I will update you when a new release is out with these fixes. Thanks for helping identifying this issue.
